### ISSUE TITLE
Make good error reporting a feature

### ISSUE THREAD

**Issue by Emoun** [Emoun/duplicate#13]
We use `proc_macro_error` to print nice errors when using `duplicate`. As #11 has shown, this is not without its risks. 

Therefore, we should make it a feature that can be disabled, so that even if build fail when using the feature, turning the feature off can resolve the issue. This feature should be enabled by default.

This could also be a win in other areas, as published crates that use `duplicate` do not need this feature (as they hopefully will have no errors.)

**Comment by CreepySkeleton** [comment]
I think `proc_macro_error` can be considered stable. I'm a biased party of course, but it's been almost a year now, and for all this time I had a real breakage only [once](https://gitlab.com/CreepySkeleton/proc-macro-error/-/issues/8), and the breakage didn't affect "normal" users who use standard rustc. 

You should also consider that maintaining a few different ways to do the same thing involves its own risks. Getting "out of sync" is the least of problems that can fall on you.

It worth notice that my crate wasn't at fault here - it was your IDE that sneaks in hidden items :) Have you reported the bug?

Frankly, I think you're overreacting.

**Comment by Emoun** [comment]
@CreepySkeleton 

I must admit the idea for this issue came when I didn't know why my build was breaking and fueled by some annoyance, however, this has subsided.

I agree with you that #11 was not the fault of `proc_macro_error` and I should report a bug to my IDE (haven't gotten to it yet). I do not consider `proc_macro_error` as unstable and this issue should not be seen as based on that.

My current justifications for this feature are as follows:

- Pretty error messages are not strictly necessary for the functioning of this crate. For users that aren't actively using `#[duplicate()]` (I.e. changing some specific invocation of it), pretty error messages are irrelevant as they have already written the invocation correctly.
- Introducing any dependency introduces complexity. This includes increasing the risk of builds failing for some reason. Even though all crates adhere to semantic versioning, there are still changes that are allowed in semantic versioning but still may be backwards incompatible. This was shown in #11.<sup>1</sup> There is value in reducing the number of hard dependencies based on this.
- Compile times are also a consideration. Say some library chooses to use this crate. They definitely won't publish a version that triggers an error in `duplicate`. Therefore, as soon as they are ready to publish, it would be wise of them to disable this feature, so that `proc_macro_error`---and with it `proc-macro-error-attr`, `proc-macro2`, `quote`, `syn`---are not unnecessarily compiled every time. Also, `duplicate` itself might run faster when it doesn't include the code for using `proc_macro_error`.
- Lastly, this is also a protection for users from me. As #11 has shown, I'm very capable of writing bad code. This gives them a way to minimize the opportunity  my mistakes affecting them.

So in conclusion: Yes this issue might have been an overreaction, but I have since thought about it with a cool head and found it still makes sense. As any good police department would do, I have investigated myself and found I did nothing wrong. 😄  

<sup>1</sup> Just to highlight how this issue is not a jab at `proc_macro_error`: Even introducing a new trait is allowed in semantic versioning but might still break builds. If users use glob imports (`use some_crate::*`), the new trait's method names might clash with another methods already in scope.

**Comment by Emoun** [comment]
Implemented and released in v0.2.6
